Add a Node/CiliumNode test based on control-plane framework

pippolo84 opened the pull request:
This PR leverages the control-plane test framework for a unit test equivalent to the `test/k8s/Node.go` Ginkgo test. With the new control plane test framework, we don't need to spin up a cluster every time the test is run.

The input is generated by the `generate.sh` script that spins up a kind cluster and manipulates the Node labels. At each label manipulation operation, the Node objects status is saved in a yaml file to be used as input for the test steps. This is needed only the first time to generate the data to feed the test.
The test will verify that, at each step, the status of the CiliumNodes, managed by the agent, is updated accordingly to the current Nodes status.

To allow the test to verify such status, a proxy is passed to the validator. The proxy exposes an API to get the current status of the k8s objects.

pippolo84 committed:
test/control-plane: add test for CiliumNodes labels

This test is a porting of a previous integration test, based on Ginkgo,
to the control-plane test framework.

The purpose of this test is to check that any update to the labels of a
k8s Node object is reflected in the corresponding CiliumNode one. The
test input is generated by the `generate.sh` script that spins up a kind
cluster and manipulates the Node labels to let Cilium updates the
CiliumNode objects.

To let the test checks both the Nodes and the CiliumNodes status, a
proxy is passed to the validator. The proxy exposes an API to get the
current status of the k8s objects in the underlying control-plane test
framework.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

nebril commented on test/control-plane/node/ciliumnodes/generate.sh:
```diff
@@ -0,0 +1,46 @@
+    kind create cluster --config manifests/kind-config-${version}.yaml --name cilium-nodes
+
+    : Install cilium
+    cilium install --wait
```
Should we specify Cilium version that is installed here, or do we expect `generate.sh` to be run in a controlled environment with latest `cilium-cli` installed?

pippolo84 replied:
Given that this script should only be run when we need to regenerate the test data, I _think_ we can assume that it will be executed in an environment with the latest version of `cilium-cli` installed.

nebril reviewed (commented):
LGTM overall, one question left inline.

aanm commented on test/control-plane/node/ciliumnodes/generate.sh:
```diff
@@ -0,0 +1,46 @@
+
+export KUBECONFIG=kubeconfig
+
+versions=(1.24)
```
Any reason why we don't have all supported versions here already?

pippolo84 replied:
No specific reason, I just thought that the latest version was enough for this test. But I can add 1.20 and 1.22 as well, like the other control-plane tests.

aanm reviewed (changes requested):
Out of curiosity, and to see the potential of this new framework, let's assume we want to extend this test. Can you add a new commit that extends the test to test the modification of a value of an existing label? Thank you

pippolo84 committed:
test/control-plane: Extend CiliumNodes labels test

Extend the CiliumNodes labels unit test to verify that, after an
overwrite of an existing label in the Node object, the agent updates the
corresponding CiliumNode object accordingly.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki reviewed (approved):
Thanks, looks good!

This points out that the API for writing the framework tests, e.g. in terms of TestSteps and validator is pretty inconvenient and a more imperative approach would be nicer (e.g. `test.UpdateObjects(...); <validate some state>`). I'd propose we'll merge this in first and then do a pass to refactor the framework. I already went few steps towards this direction in https://github.com/joamaki/cilium/blob/pr/joamaki/wip-control-plane-endpoints/test/control-plane/endpoint/endpoint_test.go. 

For a later PR I'd also want to consider using the reactor callbacks to "record" the changes made by Cilium to K8s objects rather than polling the ObjectTracker. Preferably we'd have most of the tests written in a non-polling way.

sayboras reviewed (approved):
LGTM 💯

gandro commented on test/control-plane/node/ciliumnodes/v1.20/init.yaml:
```diff
@@ -0,0 +1,341 @@
+apiVersion: v1
```
Is there a way we could mark those files as auto-generated in GitHub? 

https://docs.github.com/en/repositories/working-with-files/managing-files/customizing-how-changed-files-appear-on-github

gandro commented on test/control-plane/node/ciliumnodes/ciliumnodes_validation.go:
```diff
@@ -0,0 +1,117 @@
+	fmt.Sscanf(path.Base(stateFile), "state%d.yaml", &step)
+
+	switch step {
+	case 1:
```
nit: I personally would prefer a descriptive string here as the "step" identifier, but that's a personal preference and not a blocking comment.

joamaki replied:
I'm following up soon with a refactoring of the test framework that will do away with this.

joamaki replied:
E.g. the tests will look more like this: https://github.com/joamaki/cilium/commit/043f2cf66ab84cc40cbbb1a50b6a143c4424708b#diff-d3ea0b01ee85e6abd9fcd62e5b933d2c977f8a32fe58213e075486d922571dca

gandro reviewed (approved):
Awesome work! Two very small nits
